feat(networking): only consider close records in price setting

joshuef opened the pull request:
## Description

reviewpad:summary

joshuef committed:
feat(networking): only consider close records in price setting

joshuef commented:
I understand now @maqi  why you were so hesitant to have the `relevant` check.. As we actually filter for relevance on inbound replicated data now (or at payment time). 
 
 Hmmm... Need to chew on cache + backup implications of this approach some.

joshuef commented:
Removing do not merge, as I think we need this either way. 

As churn happens, our area of responsibility shrinks, so the price needs to reflect that, not be based upon what we're holding. 

I know we only replicate in relevant data now, but as churn happens, some of that will become "out of scope". So we need to discard based upon that. 

We can memoise the fn for every X checks or so perhaps if perf is an issue.

RolandSherwin commented on sn_networking/src/record_store.rs:
```diff
@@ -384,6 +391,14 @@ impl NodeRecordStore {
+
+        // set's the local distance range to be further if the new record is further
+        // This assumes the verified record is to be held by _us_...
+        self.distance_range = self.distance_range.map(|range| range.max(new_distance));
```
As the network grows, would we not be getting records that are closer to us? So this would mean that we'll be hanging on to the older value?

joshuef replied:
Aye, cleaned that out and moved farthest record check into an interval setter. Should be enough and prevent too much overzealous "close" checking on store etc.

joshuef committed:
feat(networking): periodically determine farthest relevant record

joshuef committed:
feat(networking): remove distance range set on put

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -428,27 +428,41 @@ impl NodeRecordStore {
-            live_time: self.timestamp.elapsed().as_secs(),
-        };
+        let live_time = self.timestamp.elapsed().as_secs();
+        let mut records_stored = self.config.max_records;
```
prefer a default vaule of `0`, to avoid in case of zero cost and no distance set,
the metrics showing `stored_records` as the max_records.

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -428,27 +428,41 @@ impl NodeRecordStore {
+            NanoTokens::zero()
+        } else if let Some(distance_range) = self.distance_range {
+            let record_keys_as_hashset: HashSet<&Key> = self.records.keys().collect();
+            let relevant_records =
```
note that there is `pruning` to be carried out when hit max_records, which is regarding ALL.
i.e. this means `relevant_records` may way below max_records, and will never reach that bar.

It may be good, may be bad,

Also, I'd suggest just rename `QuotingMetrics::records_stored` to `QuotingMetrics::relevant_records` to avoid any potential confusion.

maqi commented on sn_networking/src/record_store.rs:
```diff
@@ -428,27 +428,41 @@ impl NodeRecordStore {
+        } else if let Some(distance_range) = self.distance_range {
+            let record_keys_as_hashset: HashSet<&Key> = self.records.keys().collect();
+            let relevant_records =
+                self.get_records_within_distance_range(record_keys_as_hashset, distance_range);
```
looks like the correspondent `addr_distribution_sim` test not got updated to reflect this change?
this will result in an un-sync between the sim and real code.

joshuef replied:
Hmmm, im not sure what needs to be changed here. 

I've updated some naming, and made things a bit clearer, but the crux of the test is entirely the same. We were already only calculating for close_records as there's no churn here.

joshuef committed:
chore(networking): quoting_metrics.records_stored -> close_records_stored

joshuef committed:
test(networking): update address_sim naming for close records

maqi commented on sn_networking/src/driver.rs:
```diff
@@ -608,6 +615,21 @@ impl SwarmDriver {
+                    .kademlia
+                    .get_closest_local_peers(&self.self_peer_id.into())
+                    .map(|peer| peer.into_preimage())
+                    .take(CLOSE_GROUP_SIZE)
```
`is_in_close_range` will always return true when the input list is less than REPLICATION_RANGE (i.e. 7)
but here there is only CLOSE_GROUP_SIZE (5) peers to be passed down ?

joshuef committed:
fix(networking): get closest k peers for relevant record check

joshuef committed:
feat(networking): use an estimate of responsible distance range

uses CLOSE_GROUP_SIZE + 1 peers' distance as a proxy. This gives some margin for error
and should err on the side of caution

maqi committed:
fix(node): disable records_stored bad quote detection due to new work
